src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -610,34 +654,111 @@ private void PersistedNodeRecorder(TreePath treePath, Hash256 address, TrieNode
+    private void PruneCache()
     {
-        if (_logger.IsDebug) _logger.Debug($"Pruning nodes {MemoryUsedByDirtyCache / 1.MB()} MB , last persisted block: {LastPersistedBlockNumber} current: {LatestCommittedBlockNumber}.");
+        if (_logger.IsDebug) _logger.Debug($"Pruning nodes {DirtyMemoryUsedByDirtyCache / 1.MB()} MB , last persisted block: {LastPersistedBlockNumber} current: {LatestCommittedBlockNumber}.");
         long start = Stopwatch.GetTimestamp();
 
-        long newMemory = 0;
-
         for (int index = 0; index < _dirtyNodes.Length; index++)
         {
-            TrieStoreDirtyNodesCache dirtyNode = _dirtyNodes[index];
-            _dirtyNodesTasks[index] = Task.Run(() =>
+            int closureIndex = index;
+            TrieStoreDirtyNodesCache dirtyNode = _dirtyNodes[closureIndex];
+            _dirtyNodesTasks[closureIndex] = Task.Run(() =>
             {
-                long shardSize = dirtyNode.PruneCache(skipRecalculateMemory);
-                Interlocked.Add(ref newMemory, shardSize);
+                ConcurrentDictionary<HashAndTinyPath, Hash256?>? persistedHashes = null;
+                if (_persistedHashes.Length > 0)
+                {
+                    persistedHashes = _persistedHashes[closureIndex];
+                }
+
+                dirtyNode
+                    .PruneCache(prunePersisted: false,
+                        persistedHashes: persistedHashes,
+                        nodeStorage: _nodeStorage);
+                persistedHashes?.NoResizeClear();
             });
         }
 
         Task.WaitAll(_dirtyNodesTasks);
 
-        if (!skipRecalculateMemory) MemoryUsedByDirtyCache = newMemory;
-        _ = CachedNodesCount; // Setter also update the count
+        RecalculateTotalMemoryUsage();
 
-        if (_logger.IsDebug) _logger.Debug($"Finished pruning nodes in {(long)Stopwatch.GetElapsedTime(start).TotalMilliseconds}ms {MemoryUsedByDirtyCache / 1.MB()} MB, last persisted block: {LastPersistedBlockNumber} current: {LatestCommittedBlockNumber}.");
+        if (_logger.IsDebug) _logger.Debug($"Finished pruning nodes in {(long)Stopwatch.GetElapsedTime(start).TotalMilliseconds}ms {DirtyMemoryUsedByDirtyCache / 1.MB()} MB, last persisted block: {LastPersistedBlockNumber} current: {LatestCommittedBlockNumber}.");
+    }
+
+    /// <summary>
+    /// Only prune persisted nodes. This method attempt to pick only some shard for pruning.
+    /// </summary>
+    private void PrunePersistedNodes()
+    {
+        try
+        {
+            lock (_dirtyNodesLock)
+            {
+                long targetPruneMemory = (long)(PersistedMemoryUsedByDirtyCache * _pruningStrategy.PrunePersistedNodePortion);
+                targetPruneMemory = Math.Max(targetPruneMemory, _pruningStrategy.PrunePersistedNodeMinimumTarget);
+
+                int shardCountToPrune = (int)((targetPruneMemory / (double)PersistedMemoryUsedByDirtyCache) * 256);
```

Comment:
256 should be shard count

Reply:
Make it `const` of that name?

Reply:
There is one, I'm just reminding myself to replace this.